### ISSUE TITLE
LUA txn logging functions cause double-logging

### ISSUE THREAD

**Issue by Tristan971** [haproxy/haproxy#2315]
### Detailed Description of the Problem

When using txn:Info and similar from a LUA script, HAProxy logs the message twice

### Expected Behavior

HAProxy logs the message only once

### Steps to Reproduce the Behavior

1. Run the config in the issue
2. Look at the logs upon handling a request

### What is your configuration?

foo.lua
```lua
function foo(txn)
  txn:Info("foo" .. txn.sf:unique_id())
end

core.register_action("foo", { "http-req" }, foo)
```

haproxy.cfg
```haproxy
global
    daemon
    user haproxy
    group haproxy
    log stdout format raw local0

    lua-load-per-thread /usr/local/etc/haproxy/foo.lua

defaults simple
    mode http

    unique-id-format %[uuid()]
    unique-id-header X-Request-ID

    log global
    option httplog

    timeout check           5s
    timeout client          5s
    timeout client-fin      5ms
    timeout connect         3s
    timeout http-keep-alive 1s
    timeout http-request    3s
    timeout queue           3s
    timeout server          30s
    timeout server-fin      1s
    timeout tarpit          5s
    timeout tunnel          1h

resolvers hostdns
    parse-resolv-conf
    resolve_retries 3
    timeout resolve 1s
    timeout retry   1s

frontend http
    bind *:8080

    http-request lua.foo
    http-request return status 200 content-type text/plain lf-string "request id is: %[unique-id]"
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.8.3-86e043a 2023/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.3.html
Running on: Linux 5.15.123.1-microsoft-standard-WSL2 #1 SMP Mon Aug 7 19:01:48 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PTHREAD_EMULATION=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX +PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=32).
Built with OpenSSL version : OpenSSL 3.0.11 19 Sep 2023
Running on OpenSSL version : OpenSSL 3.0.11 19 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
[NOTICE]   (1) : New worker (8) forked
[NOTICE]   (1) : Loading success.
fooa0079ff9-50ec-46f8-b261-01d1577dafdf                         // not expected
[info] 289/115241 (8) : fooa0079ff9-50ec-46f8-b261-01d1577dafdf // expected
172.17.0.1:48218 [17/Oct/2023:11:52:41.368] http http/<NOSRV> 0/-1/-1/-1/0 200 123 - - LR-- 1/1/0/0/0 0/0 "GET / HTTP/1.1"
```

### Additional Information

Ran it using last haproxytech/haproxy-debian:2.8 docker image at the time of writing, but it happens on my own builds of HAProxy as well

An interesting aspect is that the duplicates are grouped. ie you get all unprefixed logs, then all prefixed logs in a row:
```plain
2023-10-17 12:40:49.226	blabla1
2023-10-17 12:40:49.226	blabla2
2023-10-17 12:40:49.226	blabla3
2023-10-17 12:40:49.226	blabla4
2023-10-17 12:40:49.226	[info] 289/114049 (100338) : blabla1
2023-10-17 12:40:49.226	[info] 289/114049 (100338) : blabla2
2023-10-17 12:40:49.226	[info] 289/114049 (100338) : blabla3
2023-10-17 12:40:49.226	[info] 289/114049 (100338) : blabla4
```

**Comment by Darlelet** [comment]
Indeed, it is intended (although I can't tell you what was the original intent): all messages reported through the `SEND_ERR()` macro in Lua implementation or sent through the end user lua logging functions are written to stderr in addition to the `send_log()` 

This is the historical behavior, see: c798b5d282 and 23bc375c59

It is also mentioned in the Lua documentation

> The log is sent, according with the HAProxy configuration file, on the default syslog server if it is configured and on the stderr if it is allowed.

(from: https://www.arpalert.org/src/haproxy-lua-api/2.9/index.html#TXN.log)

It can be worked around by enabling global QUIET mode, but I understand that this might be insufficiently documented / lacking configuration options.

**Comment by Darlelet** [comment]
```
fooa0079ff9-50ec-46f8-b261-01d1577dafdf                         // not expected
[info] 289/115241 (8) : fooa0079ff9-50ec-46f8-b261-01d1577dafdf // expected
```

First line (raw message without header), is indirectly generated by the `send_log()` call, according to your global log configuration (`log stdout format raw local0`) in the global section, the second one is generated from Lua implementation and written to stderr.

**Comment by Tristan971** [comment]
> It is also mentioned in the Lua documentation
>> The log is sent, according with the HAProxy configuration file, on the default syslog server if it is configured and on the stderr if it is allowed.
>>
> (from: https://www.arpalert.org/src/haproxy-lua-api/2.9/index.html#TXN.log)

Right, I missed that bit...

That is quite surprising behaviour to be honest...

And in practice I actually get **3** readings, to be more exact:
- 2 on my log (tag "haproxy" below, ie syslog app="haproxy") - and yes I have only 1 configured
- 1 on stderr (tag "haproxy.service" below, ie systemd unit="haproxy.service")

![2023-10-17 13_18_24](https://github.com/haproxy/haproxy/assets/875533/7d34c5d0-2609-4073-9398-dc0df8ea24b3)

Finally, since `quiet` documentation reads:

> Do not display any message during startup. It is equivalent to the command-
line argument "-q".

and thus doesn't mention this, I'm a bit hesitant to enable it as I don't know what other messages it would end up suppressing...

**Comment by Tristan971** [comment]
The extra third log is probably because `option log-separate-errors` affects it somehow as well...

This all makes lua logging a bit weird tbh

Relevant config portions:
```haproxy
global
...
    log /dev/log len 8192 local0
...

default ...
    log global
    option log-separate-errors
    log-format       "..."
    error-log-format "..."
```

**Comment by Tristan971** [comment]
Well, a few more findings:
1. `quiet` in global changes nothing (add it to the issue's sample configuration)
2. `option log-separate-errors` isn't the reason for the duped syslog message on my end, so the search on that side continues...

**Comment by Darlelet** [comment]
It seems like the third message is not generated from Lua but from haproxy itself (during server selection):
```172.17.0.1:48218 [17/Oct/2023:11:52:41.368] http http/<NOSRV> 0/-1/-1/-1/0 200 123 - - LR-- 1/1/0/0/0 0/0 "GET / HTTP/1.1"```

Can't correlate that with the screenshot you provided because of the blurred infos, but the last report suggests that the message has a greater length.

About the QUIET mode workaround:
>and thus doesn't mention this, I'm a bit hesitant to enable it as I don't know what other messages it would end up suppressing...

~I agree with you, it was probably done that way at that time for some reasons, but it pretty much looks like a hack now.~
-> Edit: see below

>This all makes lua logging a bit weird tbh

Yeah there's probably room for improvement  since this code has not evolved since the initial implementation indeed.

**Comment by Tristan971** [comment]
> It seems like the third message is not generated from Lua but from haproxy itself (during server selection):
>> 172.17.0.1:48218 [17/Oct/2023:11:52:41.368] http http/<NOSRV> 0/-1/-1/-1/0 200 123 - - LR-- 1/1/0/0/0 0/0 "GET / HTTP/1.1"

Third message in my logs isn't the HTTP log no. Here's a less agressively redacted sample 😅 

![2023-10-17 13_39_16](https://github.com/haproxy/haproxy/assets/875533/573905d1-58cc-41d0-aa79-d3bc28932aad)

> I agree with you, it was probably done that way at that time for some reasons, but it pretty much looks like a hack now.

Since it also doesn't work, at least we don't have to worry about that anymore :p

**Comment by Darlelet** [comment]
>Third message in my logs isn't the HTTP log no. Here's a less agressively redacted sample 😅

Thanks! Then maybe it's systemd duplicating stderr messages only? Because only the prefixed one is getting reported to stderr (the "superfluous" one in your case, generated from Lua implementation), the other one is reported to stdout according to your `log` directive.

**Comment by Tristan971** [comment]
The way systemd logs work is that it captures stdout/stderr of the process it manages, and forwards that to the system log (ie systemd-journal, what `journalctl -u haproxy` reads from).

So it is completely normal that anything hitting stdout/stderr is caught by systemd. That makes sense to me.

> the other one is reported to stdout according to your log directive

Nope where my actual config in https://github.com/haproxy/haproxy/issues/2315#issuecomment-1766308310

```haproxy
global
...
    log /dev/log len 8192 local0
...

default ...
    log global
    option log-separate-errors
    log-format       "..."
    error-log-format "..."
```

I don't have stdout logging in prod (I would prefer it than a dedicated syslog setup but alas systemd-logging is one really slow piece of software)

So with this I expect:
- 1 message from stderr (caught by systemd) => I see it
- 1 message from syslog (caught by the sink at /dev/log) => I see 2 of these

**Comment by Tristan971** [comment]
Here's a screenshot that summarizes a bit better, according the comments above. (note that this is *with* `quiet`)

![2023-10-17 13_53_08](https://github.com/haproxy/haproxy/assets/875533/69f140a6-97ef-46e5-bce4-5b3bbb9c3ff0)

**Comment by Tristan971** [comment]
Well, after much testing including using a syslog sink locally and all, I can't repro my extra duping... I'll keep digging but let's ignore it for now.

I guess that leaves only `quiet` not working as a (reproducible) bug. Which is meh.

I suppose this should be converted into a feature request to treat LUA logs like any other log and send it only once (to the configured log target).

**Comment by Darlelet** [comment]
Thanks, so I suspect that somehow the message written to `stderr` gets incorrectly reported as coming from your sink (or is being relayed somehow), because I don't see how the Lua-generated message (with the prefix), can end up being reported to `/dev/log` (through the configured `log` destination) with just haproxy involved given that it is a direct call to `fprintf(stderr...)`

Here is what I suspect:
```
    txn:log()  -> log through /dev/log according to your `log` settings -> haproxy EU sink
               -> stderr -> haproxy EU sink
                         -> systemd-journal
```

Now for the QUIET mode, it does work on my end and what it does is that it completely suppresses CLI messages (including the Lua generated ones), it seems that the check in Lua code is not even necessary given that haproxy redirects messages to `/dev/null` when QUIET is set. Note that in this case log messages will still be sent to proper log destinations (unless `stderr` or `stdout` are used as log destination)

**Comment by Tristan971** [comment]
> Thanks, so I suspect that somehow the message written to stderr gets incorrectly reported as coming from your sink (or is being relayed somehow)

You must be right, though that is rather confusing to me... Oh well.

> Now for the QUIET mode, it does work on my end

Well, I can still reproduce logs there with a non-stdout log configured. Unzip + docker-compose up + curl localhost:8080 and see for yourself I guess :/ [test.zip](https://github.com/haproxy/haproxy/files/12929389/test.zip)

![2023-10-17 14_26_01](https://github.com/haproxy/haproxy/assets/875533/f89a70a9-3da5-41c4-95fb-bf136dce1a58)

**Comment by Darlelet** [comment]
Oh, maybe QUIET mode is conflicting with some options that were used to start haproxy on your side (just guessing because your test setup is involving docker vs manually started haproxy), I'm starting it with `haproxy -f repro.conf`

But indeed, I don't think we should be solely relying on QUIET mode to suppress duplicated messages from Lua, maybe we should add a specific tunable and / or change default behavior, and in that sense I think you're right that we may need to discuss that in a dedicated feature request :)

**Comment by Darlelet** [comment]
Actually it's the `-db` option which makes QUIET mode ineffective (`-db` sets MODE_FOREGROUND on)

```c
        if (arg_mode & (MODE_DEBUG | MODE_FOREGROUND)) {
                /* command line debug mode inhibits configuration mode */
                global.mode &= ~(MODE_DAEMON | MODE_QUIET);
                global.mode |= (arg_mode & (MODE_DEBUG | MODE_FOREGROUND));
        }

```

**Comment by Tristan971** [comment]
fwiw tweaking how I handle logging, ie going from `haproxy -> rsyslog -> promtail` to `haproxy -> promtail` did remove the extra dupe, so that one was indeed on my end.

---

> Actually it's the -db option which makes QUIET mode ineffective (-db sets MODE_FOREGROUND on)

That's a fair possibility. Not only does `db` defeat it, but one of these also does then (excerpt from ps ax; as I don't use docker for it in prod and only provided it as self-contained repro):
```
/usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -dMno-merge,cold-first,caller
/usr/sbin/haproxy -sf 101874 -x sockpair@5 -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -dMno-merge,cold-first,caller
```

> But indeed, I don't think we should be solely relying on QUIET mode to suppress duplicated messages from Lua, maybe we should add a specific tunable and / or change default behavior, and in that sense I think you're right that we may need to discuss that in a dedicated feature request :)

Do you mean I should open another completely separate github issue? Figured converting this one would make more sense but maybe it's become too noisy... hmm...

**Comment by Darlelet** [comment]
Glad to hear that you found the cause for the dup' logs!


>That's a fair possibility. Not only does db defeat it, but one of these also does then (excerpt from ps ax; as I don't use docker for it in prod and only provided it as self-contained repro):

Nice catch! Indeed the `-Ws` (to enable systemd support for master-worker) also sets FOREGROUND mode
```c
 else if (*flag == 'W' && flag[1] == 's') {
                                arg_mode |= MODE_MWORKER | MODE_FOREGROUND;
#if defined(USE_SYSTEMD)
```

>Do you mean I should open another completely separate github issue? Figured converting this one would make more sense but maybe it's become too noisy... hmm...

Didn't even know this was a possibility.. as you wish then :smile:

**Comment by Tristan971** [comment]
> Indeed the -Ws (to enable systemd support for master-worker) also sets FOREGROUND mode

right, so most cases can't use it then haha. Well, it was not a good solution anyway, so not a big loss.

> Didn't even know this was a possibility.. as you wish then 😄

Right... Hmm, I guess it will be cleaner in a new one without the debugging bits, and we can cross-link for those wanting context.

**Comment by Tristan971** [comment]
Fyi if someone runs into this eventually, this is the "clever" reason I was getting dupes earlier 🔫 

```
$ cat /etc/systemd/journald.conf | grep ForwardToSyslog
#ForwardToSyslog=yes
```

**Comment by Tristan971** [comment]
After more tests, the situation is even weirder! With this:

```
global
  ...
  log stderr format raw local7
  ...

ring syslog-sink
    ....
    server sink 10.1.1.11:514

defaults simple
    ...
    log ring@syslog-sink local0
```

then I get:
```
# stderr
test-haproxy-1      | [info] 289/154633 (8) : foo881dbfa5-bd9e-4716-a175-f345e9f29ebf
# syslog
test-syslog-sink-1  | ... - haproxy 8 - - foo881dbfa5-bd9e-4716-a175-f345e9f29ebf
test-syslog-sink-1  | ... - haproxy 8 - - src=10.1.1.1 [...] request="GET / HTTP/1.1"
```

so you get the good log sent to the wrong place (stderr), and the bad one to actual log configured... fun :)

**Comment by Darlelet** [comment]
Hmm no actually it's the expected behavior, you inverted the bad and the good one in your example in the new issue I think

(the stderr one is the one prefixed with some header data, the good one is formatted according to your `log` directive parameters)

```
        send_log(px, level, "%s\n", trash.area); // good one
        if (!(global.mode & MODE_QUIET) || (global.mode & (MODE_VERBOSE | MODE_STARTING))) {
                if (level == LOG_DEBUG && !(global.mode & MODE_DEBUG))
                        return;

                get_localtime(date.tv_sec, &tm);
                fprintf(stderr, "[%s] %03d/%02d%02d%02d (%d) : %s\n",
                        log_levels[level], tm.tm_yday, tm.tm_hour, tm.tm_min, tm.tm_sec,
                        (int)getpid(), trash.area); // duplicated on stderr (with prefix msg)
                fflush(stderr);
        }
```

**Comment by Tristan971** [comment]
Hmmm... alright. Somehow had understood the opposite but fair enough. 🤔 

In this situation I can probably provide a patch actually.

**Comment by Darlelet** [comment]
My bad if I wasn't clear enough haha

As for why you saw a change of behavior in your last test when switching to intermediate ring instead of UDP output: the default log format in ring sections is `raw` (and it overrides the log format from log directive which is `local` by default), but it can be specified manually to other values through the `format` directive from the ring backend directly.

> In this situation I can probably provide a patch actually.

Yeah sure! Thanks in advance for that and feel free to ask for hints if needed! 😄

**Comment by Tristan971** [comment]
> the default log format in ring sections is raw (and it overrides the log format from log directive which is local by default), but it can be specified manually to other values through the format directive from the ring backend directly.

right, makes sense indeed

> feel free to ask for hints if needed! 😄

Well, took a stab at it; not sure of the approach for the config setting but the rest is hopefully as straightforward as it seemed to me :)

**Comment by Darlelet** [comment]
>not sure of the approach for the config setting

Hmm, how about registering a  new dedicated `tune.lua` keyword to configure the logging behavior (ie: toggler to enable/disable stderr logging)?

https://github.com/haproxy/haproxy/blob/4070e4042a1fbe64fa820992c63af7284735e854/src/hlua.c#L12679

Edit: forget about this comment, didn't notice that you already sent a patch to the ML :)
